plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/IncorrectDateTimeFormatInspection.java

```diff
@@ -0,0 +1,262 @@
+      }
+
+      private void checkInQuoteProblem(@NotNull PsiExpression expression, boolean inQuote, int lastQuoteIndex) {
+        if (inQuote) {
+          TextRange range = ExpressionUtils.findStringLiteralRange(expression, lastQuoteIndex, lastQuoteIndex + 1);
+          if (range != null) {
+            holder.registerProblem(expression, range,
+                                   InspectionGadgetsBundle.message("inspection.incorrect.date.format.message.unpaired", '\''),
+                                   LocalQuickFix.EMPTY_ARRAY);
+          }
+        }
+      }
+
+      private void checkUnsupportedSymbolsProblem(@NotNull PsiExpression expression, int pos, char c) {
+        if (c == '{' || c == '}' || c == '#') {
+          TextRange range = ExpressionUtils.findStringLiteralRange(expression, pos, pos + 1);
+          holder.registerProblem(expression, range,
+                                 InspectionGadgetsBundle.message("inspection.incorrect.date.format.message.unsupported", c),
+                                 LocalQuickFix.EMPTY_ARRAY);
+        }
+      }
+
+      private void checkPadding(@NotNull PsiExpression expression, char[] array, int firstLetterIndex, int nextAfterP) {
+        if (array[firstLetterIndex] == 'p') {
+          if (nextAfterP >= array.length || !isLetter(array[nextAfterP])) {
+            TextRange range = ExpressionUtils.findStringLiteralRange(expression, firstLetterIndex, nextAfterP);
+            holder.registerProblem(expression, range,
+                                   InspectionGadgetsBundle.message("inspection.incorrect.date.format.message.padding",
+                                                                   array[firstLetterIndex]),
+                                   LocalQuickFix.EMPTY_ARRAY);
+          }
+        }
+      }
+
+      private boolean isLetter(char c) {
+        return c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z';
+      }
+
+      private void checkAvailableCount(PsiExpression expression, @Nullable Token token, ProblemsHolder holder) {
+        if (token == null) {
+          return;
+        }
+        if (token.character == 'p') {
+          //check during tokenizing
+          return;
+        }
+        CountVerifier verifier = DATE_TIME_FORMATTER_ALLOWED.get(token.character);
+        if (verifier == null || !verifier.verifier(token.length)) {
+          TextRange range = ExpressionUtils.findStringLiteralRange(expression, token.pos, token.pos + token.length);
+          holder.registerProblem(expression, range,
+                                 InspectionGadgetsBundle.message("inspection.incorrect.date.format.message.unsupported",
+                                                                 token.toString()),
+                                 LocalQuickFix.EMPTY_ARRAY);
+        }
+      }
+    };
+  }
+
+  private static class Token {
+    final char character;
+    final int pos;
+    final int length;
+
+    Token(int pos, char[] chars, int length) {
```

Comment:
Probably reorder parameters and put chars at the first place? It's commonly used this way in method signatures (e.g., `java.io.Reader#read(char[], int, int)`)

Reply:
Parameters were reordered. Thank you!